substratevm/src/com.oracle.svm.core/src/com/oracle/svm/core/jfr/oldobject/OldObjectSampler.java

```diff
@@ -0,0 +1,140 @@
+ * or visit www.oracle.com if you need additional information or have any
+ * questions.
+ */
+
+package com.oracle.svm.core.jfr.oldobject;
+
+import com.oracle.svm.core.Uninterruptible;
+import com.oracle.svm.core.thread.JavaThreads;
+
+import java.lang.ref.WeakReference;
+
+final class OldObjectSampler {
+    private final OldObjectArray samples;
+    private final OldObjectPriorityQueue queue;
+    private final OldObjectList list;
+    private final OldObjectEffects effects;
+    private long totalAllocated;
+
+    OldObjectSampler(int queueSize, OldObjectList list, OldObjectEffects effects) {
+        this.samples = new OldObjectArray(queueSize);
+        this.queue = new OldObjectPriorityQueue(this.samples);
+        this.list = list;
+        this.effects = effects;
+    }
+
+    @Uninterruptible(reason = "Accesses allocation profiler.")
+    void sample(WeakReference<Object> ref, long allocatedSize, int arrayLength) {
+        totalAllocated += allocatedSize;
+        long span = totalAllocated - queue.total();
+        if (queue.isFull()) {
+            if (queue.peek().span > span) {
+                // Sample will not fit, try to scavenge
+                // HotSpot code scavenges when it gets notified that a sample was GC'd,
+                // but such mechanism doesn't exist in substratevm.
+                // So, instead just attempt to scavenge when the queue is full.
+                int numDead = scavenge();
+                if (numDead == 0) {
+                    // Sample will not fit and all objects still in use, return early
+                    return;
+                }
+            } else {
+                // Offered element has a higher priority,
+                // vacate from the lowest priority one and insert the element.
+                evict();
+            }
+        }
+
+        store(ref, allocatedSize, effects.elapsedTicks(), arrayLength);
+    }
+
+    @Uninterruptible(reason = "Accesses allocation profiler.")
+    private int scavenge() {
+        int numDead = 0;
+        OldObject current = list.head();
+        while (current != null) {
+            OldObject next = current.previous;
+            final WeakReference<?> ref = current.reference;
+            if (ref.get() == null) {
+                remove(current);
+                numDead++;
+            }
+
+            current = next;
+        }
+        return numDead;
+    }
+
+    /**
+     * Remove a given sample from the sampler.
+     */
+    @Uninterruptible(reason = "Accesses allocation profiler.")
+    private void remove(OldObject sample) {
+        final OldObject prev = sample.previous;
+        if (prev != null) {
+            // To keep samples evenly distributed over time,
+            // the weight of a sample that is removed should be redistributed.
+            // Here it gets redistributed to the sample that came just after in time.
+            queue.remove(prev);
+            prev.span += sample.span;
```

Comment:
Another interesting observation which also needs some further explanation :)

Because the span is added it to the younger neighbour, that is why I implemented it as a single-linked list. In HotSpot, this is implemented as double-linked list so it could efficiently distribute the span to the older sample. I went for the simpler, more space efficient, option, taking into account that the default sample size is relatively small, so when removing things I can just loop around and fix things up.

Now, about how to distribute the span. [Here's what Erik said recently when I asked him about the linked list of samples](https://mail.openjdk.org/pipermail/hotspot-jfr-dev/2023-June/005067.html):

> The weight of the object, the allocation span, in the time-based list, should be divided to its neighbour(s).

Then [@tstuefe pointed out](https://mail.openjdk.org/pipermail/hotspot-jfr-dev/2023-June/005069.html):

> span is redistributed to its (in time) predecessor sample. Following your explanation, should the span not be evenly divided between predecessor and successor?

The first part is not entirely correct since `prev` is really the sample that was sampled after the current object, so indeed the younger one. But the point is that Erik suggest it should be divided. So, 50% to younger and 50% to older. Unfortunately Erik didn't reply to this, so I think it would make sense to follow up on hotspot-jfr-dev and see if HotSpot should change accordingly. Feel free to kick off a discussion there since I'll be away for a bit.

Reply:
Thanks for the detailed response! Ok that all makes sense to me. I agree that a singly-linked is the best option to achieve the same behavior as in hotspot (only redistribute to the younger neighbor).

That's a good idea. I'll follow up on your "hotspot-jfr-dev" email thread asking "why doesn't the Hotspot implementation actually distribute the span evenly to both neighbors? And if it must distribute unevenly, why not choose the older neighbor?".

Reply:
I asked on the hotspot mailing list here: https://mail.openjdk.org/pipermail/hotspot-jfr-dev/2023-June/005164.html

Unfortunately, there was no response.